File: src/12-challenge/solution.ts
type Elf = {
	actions: string
	lifePoints: number
}

type Action = 'A' | 'B' | 'F'
type ActionResponse = (victimLifePoints: Elf['lifePoints'], victimCounterAction: Action) => number

export default function fn(elf1: string, elf2: string): number {
	let elf01: Elf = {
		actions: elf1,
		lifePoints: 3,
	}

	let elf02: Elf = {
		actions: elf2,
		lifePoints: 3,
	}

	const actionResponses: Record<Action | (string & {}), ActionResponse> = {
		A: (victimLifePoints, victimCounterAction) => victimLifePoints - +(victimCounterAction !== 'B'),
		B: (victimLifePoints, _) => victimLifePoints,
		F: (victimLifePoints, _) => victimLifePoints - 2,
		undefined: (victimLifePoints, _) => victimLifePoints,
	}

	const minActionsLength = Math.min(elf01.actions.length, elf02.actions.length)

	for (let actionI = 0; actionI < minActionsLength && elf01.lifePoints && elf02.lifePoints; actionI++) {
		const elf01Action = elf01.actions[actionI]
		const elf02Action = elf02.actions[actionI]

		elf02.lifePoints = actionResponses[elf01Action](elf02.lifePoints, elf02Action as Action)
		elf01.lifePoints = actionResponses[elf02Action](elf01.lifePoints, elf01Action as Action)
	}

	if (elf01.lifePoints > elf02.lifePoints) return 1
	return elf02.lifePoints > elf01.lifePoints ? 2 : 0
}

File: src/12-challenge/solution.py
from typing import TypedDict, Literal, Callable, Dict, cast

Elf = TypedDict("Elf", {"actions": str, "life_points": int})

Action = Literal["A", "B", "F"]
ActionResponse = Callable[[int, Action], int]


def fn(elf1: str, elf2: str) -> int:
    elf01: Elf = {"actions": elf1, "life_points": 3}
    elf02: Elf = {"actions": elf2, "life_points": 3}

    action_responses: Dict[Action, ActionResponse] = {
        "A": lambda victim_life_points, victim_counter_action: (
            victim_life_points
            if victim_counter_action == "B"
            else victim_life_points - 1
        ),
        "B": lambda victim_life_points, _: victim_life_points,
        "F": lambda victim_life_points, _: victim_life_points - 2,
    }

    min_actions_length = min(len(elf01["actions"]), len(elf02["actions"]))

    for action_i in range(min_actions_length):
        if not elf01["life_points"] or not elf02["life_points"]:
            break

        elf01_action = cast(Action, elf01["actions"][action_i])
        elf02_action = cast(Action, elf02["actions"][action_i])

        elf02["life_points"] = action_responses[elf01_action](
            elf02["life_points"], elf02_action
        )
        elf01["life_points"] = action_responses[elf02_action](
            elf01["life_points"], elf01_action
        )

    if elf01["life_points"] > elf02["life_points"]:
        return 1

    return 2 if elf02["life_points"] > elf01["life_points"] else 0
